### PR TITLE
Launchpad Checklist API: Add checklist and task registry

### PR DESCRIPTION
**Pull request by jeyip**
Fixes #

## Proposed changes:

* Create a centralized registry for Launchpad checklists and Launchpad checklist tasks
* By doing so, we can create easy to use helper methods ( `register_launchpad_task_list`, `register_launchpad_task` ) that will allow other developers to spin up Launchpad checklists quickly
* To elaborate, all of this will allow others to programmatically register Launchpad checklists and tasks wherever is convenient in wpcom rather than hardcoding information in the `jetpack-mu-wpcom` `launchpad.php` file

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Testing instructions:

1. Run `bin/jetpack-downloader test jetpack-mu-wpcom-plugin add/registry-for-launchpad-checklists` to build and sync this branch to your sandbox. 
2. Sandbox a test site and public API
3. Start any new launchpad onboarding site Ex. wordpress.com/setup/free/intro
4. Go to `https://developer.wordpress.com/docs/api/console/`, select Rest API and wpcom/v2, and input the following: `/sites/YOURSITESLUG/launchpad/checklist?checklist_slug=free`. Confirm you get back an appropriate checklist that matches the status of checklist items on Launchpad. 
5. Add the following code to the bottom of the `launchpad.php` file. We are going to verify that the checklist registration logic works properly by programmatically adding an entirely new checklist

```
function register_fake_checklist_and_tasks() {
	$fake_checklist = array(
		'id'          => 'fake_checklist',
		'title'       => 'Fake Checklist',
		'task_ids'       => array(
			'fake_task_1',
			'fake_task_2',
		),
	);

	register_launchpad_task_list( $fake_checklist );

	$fake_tasks = array(
		array(
			'id'        => 'fake_task_1',
			'title'     => 'Fake Task 1',
			'completed' => false,
			'disabled'  => false,
		),
		array(
			'id'        => 'fake_task_2',
			'title'     => 'Fake Task 2',
			'completed' => false,
			'disabled'  => false,
		)
		);

	register_launchpad_tasks( $fake_tasks );
}

add_action( 'init', 'register_fake_checklist_and_tasks' );
```
6. Then add `fake_checklist` as [a whitelisted query param on line 53](https://github.com/Automattic/jetpack/blob/trunk/projects/packages/jetpack-mu-wpcom/src/features/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-launchpad-checklist.php#L53) of the `class-wpcom-rest-api-v2-endpoint-launchpad-checklist.php` file
7. Verify in the developer console that `/sites/YOURSITESLUG/launchpad/checklist?checklist_slug=fake_checklist` returns a collection of fake tasks that we registered. 

Screenshot 2023-04-20 at 7 01 18 PM

### CODE REVIEW

File: projects/packages/jetpack-mu-wpcom/src/class-jetpack-mu-wpcom.php
Comment (mattwiebe) [inline projects/packages/jetpack-mu-wpcom/src/class-jetpack-mu-wpcom.php:36]: Once we want to centralize things in our plugin and unload the existing logic on wpcom, we'll first move this line to an earlier priority.
```php
add_action( 'plugins_loaded', array( __CLASS__, 'load_launchpad' ), 9 );
```

This will get us in just before the default priority `10`, and tidily allow us unhook what's currently running on wpcom and Launchpad inside the `load_launchpad` method below.

```php
if ( class_exists( 'WPCOM_Launchpad' ) {
  remove_action( 'plugins_loaded', array( WPCOM_Launchpad::get_instance(), 'init' ) );
}
```

Once that's running, we can remove `WPCOM_Launchpad` from the Simple and `wpcomsh` codebases safely. (Although first we'll have to audit to ensure nothing else is calling `WPCOM_Launchpad`)

File: projects/packages/jetpack-mu-wpcom/src/features/launchpad/class-launchpad-task-lists.php
Comment (mattwiebe) [inline projects/packages/jetpack-mu-wpcom/src/features/launchpad/class-launchpad-task-lists.php]: We generally want stuff to be pretty lenient in WP(com) and not throw errors, at least never in production. There's too much stuff interacting weirdly across a large codebase like ours, so let's take a different approach. We do want devs to get feedback on how something's wrong, but not raise errors. A pretty good way is to use `_doing_it_wrong` which will send to the error log, and as [its docs](https://developer.wordpress.org/reference/functions/_doing_it_wrong/) say:

> The current behavior is to trigger a user error if `WP_DEBUG` is true.

Which is a good way to run things on your sandbox.

**Review by mattwiebe — COMMENTED** [review]
Off to a great start here, left some feedback on improvements and notes for the future

File: projects/packages/jetpack-mu-wpcom/src/features/launchpad/class-launchpad-task-lists.php
Comment (jeyip) [reply projects/packages/jetpack-mu-wpcom/src/features/launchpad/class-launchpad-task-lists.php]: > We do want devs to get feedback on how something's wrong, but not raise errors. A pretty good way is to use _doing_it_wrong which will send to the error log, and as [its docs](https://developer.wordpress.org/reference/functions/_doing_it_wrong/) say:

Great feedback! This is exactly what I was trying to accomplish by setting the `trigger_error` error level to `E_USER_WARNING` ( doesn't fatal, but leaves a message in the error log )

I'll make the changes

Comment (jeyip) [reply projects/packages/jetpack-mu-wpcom/src/features/launchpad/class-launchpad-task-lists.php]: The `_doing_it_wrong` method doesn't seem to be sending anything to the error log, even when I set `WP_DEBUG` to true in the `wp-config` of my sandbox.

Not sure if I'm doing something wrong? Wondering if you see anything obvious [6615a57](https://github.com/Automattic/jetpack/pull/30211/commits/6615a573998869e387b592b6d63569e6d38bbb5c)

Screenshot 2023-04-21 at 1 27 02 PM

File: projects/packages/jetpack-mu-wpcom/src/class-jetpack-mu-wpcom.php
Comment (jeyip) [reply projects/packages/jetpack-mu-wpcom/src/class-jetpack-mu-wpcom.php:36]: Makes sense to me 👍 

I'll make a note in our task list tracker about this release strategy

Comment (jeyip) [reply projects/packages/jetpack-mu-wpcom/src/class-jetpack-mu-wpcom.php:36]: Noted https://github.com/Automattic/wp-calypso/issues/75504#issuecomment-1518306697

File: projects/packages/jetpack-mu-wpcom/src/features/launchpad/class-launchpad-task-lists.php
Comment (jeyip) [reply projects/packages/jetpack-mu-wpcom/src/features/launchpad/class-launchpad-task-lists.php]: We can handle this in a follow up next sprint -- leaving a note in the tracking issue

Comment (jeyip) [reply projects/packages/jetpack-mu-wpcom/src/features/launchpad/class-launchpad-task-lists.php]: Noted

**Comment by agrullon95** [comment]
I'm able to verify that the changes allow me to register new checklists / tasks and retrieve the checklist information from the Launchpad checklist endpoint.

I'm not able to figure out where to see the `_doing_it_wrong` messages, but the overall validation logic is working either way. I also messed around with the task list / tasks properties to force unexpected values and the endpoint returned an empty checklist array as expected.

File: projects/packages/jetpack-mu-wpcom/src/class-jetpack-mu-wpcom.php
Comment (agrullon95) [inline projects/packages/jetpack-mu-wpcom/src/class-jetpack-mu-wpcom.php:77]: We can keep this `/features/launchpad/launchpad.php` import and eventually remove the same import from the constructor in `class-wpcom-rest-api-v2-endpoint-launchpad-checklist.php`

**Review by agrullon95 — APPROVED** [review]

File: projects/packages/jetpack-mu-wpcom/src/features/launchpad/launchpad.php
Comment (mattwiebe) [inline projects/packages/jetpack-mu-wpcom/src/features/launchpad/launchpad.php:362]: I would like to circle back and refactor this to explicitly call `register_launchpad_checklist` for each individual task and task list but I don't mind starting out in this state.

**Review by mattwiebe — APPROVED** [review]
Great work here. Nitpicks can wait for after we get this merged.

File: projects/packages/jetpack-mu-wpcom/src/features/launchpad/launchpad.php
Comment (jeyip) [reply projects/packages/jetpack-mu-wpcom/src/features/launchpad/launchpad.php:362]: Sounds good 👍

Comment (obenland) [reply projects/packages/jetpack-mu-wpcom/src/features/launchpad/launchpad.php:362]: It might also be good to [prefix function names](https://developer.wordpress.org/plugins/plugin-basics/best-practices/#prefix-everything) to avoid potential collisions on WoA sites

**Pull request merged**
